libsolidity/codegen/ExpressionCompiler.cpp

```diff
@@ -410,6 +410,38 @@ bool ExpressionCompiler::visit(TupleExpression const& _tuple)
 bool ExpressionCompiler::visit(UnaryOperation const& _unaryOperation)
 {
 	CompilerContext::LocationSetter locationSetter(m_context, _unaryOperation);
+
+	FunctionDefinition const* function = *_unaryOperation.annotation().userDefinedFunction;
+	if (function)
+	{
+		solAssert(function->isFree() || function->libraryFunction());
+
+		FunctionType const* functionType = _unaryOperation.userDefinedFunctionType();
+		solAssert(functionType);
+
+		functionType = dynamic_cast<FunctionType const&>(*functionType).withBoundFirstArgument();
```

Comment:
Why indirect this via a function with a bound first argument? I don't see any first argument being bound in any of this - I'd just call the function with the operators as arguments, is there any difference?

Reply:
Not sure. It seems to have been like this in @chriseth's initial implementation: 3bd047f1881073b2002a31baed7fb974dbc20c57. Unless that was fixed up into it later but I doubt it. In any case it's how it was in the original PR and @chriseth did not complain about it even though he paid a lot of attention to the codegen part (e.g. https://github.com/ethereum/solidity/pull/12362#discussion_r915822159). I suspect this was started as a copy-paste of the code for handling bound functions and this part was not wrong so it survived in that form even if it was not strictly necessary.

But yeah, now that you pointed it out, it seems unnecessary other than to access `selfType()` which is also just a helper to access the first argument type. I will switch to a plain function type.